### DISCUSSION TITLE
Customize the list of generated views

### DISCUSSION THREAD

**Discussion by pas256** [rails/tailwindcss-rails#425] [Q&A] [answered]
I would like to add 1 more view to get generated by `rails g scaffold` so I can have a `views/_list_item.rb` file.

If I could modify `available_views` in [`scaffold_generator.rb`](https://github.com/rails/tailwindcss-rails/blob/main/lib/generators/tailwindcss/scaffold/scaffold_generator.rb#L31) I would be able to accomplish this, but unfortunately it is private and not customizable.

I started trying to monkey patch the file by doing something like this

```ruby
require 'tailwindcss-rails/generators/tailwindcss/scaffold/scaffold_generator'

Tailwindcss::Generators::ScaffoldGenerator.available_views = %w[index edit show new _form _list_view]
```

but unlike the rails generators where I can do this for generating pundit policies
```ruby
require 'rails/generators'
require 'rails/generators/rails/scaffold/scaffold_generator'

# Invoke with scaffolding generators
Rails::Generators::ScaffoldGenerator.hook_for :policy, default: true, type: :boolean
```

this one raises a `LoadError` and I can't figure out why.

I know this isn't a discussion forum, but any guidance would be greatly appreciated. I feel dumb for not figuring this out. Thanks!

**Comment by flavorjones** [comment]
@pas256 Thanks for asking this question. I'm not sure I'll be totally helpful, but I'll try.

This gem's scaffold generator is a subclass of [`railties/lib/rails/generators/erb/scaffold/scaffold_generator.rb`](https://github.com/rails/rails/blob/main/railties/lib/rails/generators/erb/scaffold/scaffold_generator.rb) and is structured nearly identically, with a hard-coded list of available views.

Are you saying that it's possible to add a view to the basic Rails scaffold, and that approach doesn't work with this gem? Or are you asking about adding views more generally? I'm not aware of any way to do what you're asking with vanilla Rails, but I'm also not deeply familiar with the stock generators.

Edit: "I'm not aware" beyond what's suggested at https://guides.rubyonrails.org/generators.html which is to subclass the generator and set it as your app's default. Have you tried that?

**Reply by pas256** [reply, marked as answer]
Yeah it is possible. Well, sort of.

I was following this: https://gist.github.com/annikoff/331f785aa7a207a7945b1eca6eff526b to add pundit policy generation to my scaffolding. The technique it uses is to make a `hooks.rb` file that does the monkey patching.

... however the approach of `require`ing inside of the generator config in `config/application.rb` felt weird, and ended up causing the `active_record` generator to be skipped. `¯\_(ツ)_/¯`

This morning I have abandon that approach and went back to subclassing as you suggested. I have a working solution, and I am mostly happy with it, but it does have more duplicated code than is strictly necessary.

First, override the `scaffold_controller` generator:

```ruby
# config/application.rb

config.generators do |g|
  # Use my custom scaffold controller generator
  g.scaffold_controller :custom_scaffold_controller
  g.fallbacks[:custom_scaffold_controller] = :scaffold_controller
end
```

which points to this custom generator. This technique of subclassing and adding only the hooks necessary seemed like a better approach to me:

```ruby
# lib/generators/rails/custom_scaffold_controller/custom_scaffold_controller_generator.rb

module Rails
  # My custom generator for the scaffold controller that includes the policy generation
  # on top of the usual Rails scaffold controller generator.
  class CustomScaffoldControllerGenerator < Rails::Generators::ScaffoldControllerGenerator
    # Add Pundit policy generation
    hook_for :policy, default: true, type: :boolean

    # Add i18n locale generation for models and views
    hook_for :locale, default: true, type: :boolean

    # Add extra views for the scaffold
    hook_for :views, default: true, type: :boolean
  end
end
```

For that last hook, I would like to have:

```ruby
# lib/generators/rails/views/views_generator.rb

require "tailwindcss-rails/generators/tailwindcss/scaffold/scaffold_generator"

module Rails
  module Generators
    # Generate custom views
    class ViewsGenerator < Tailwindcss::Generators::ScaffoldGenerator

      def available_views
        %w(index edit show new _form _list_view)
        # ...or use super and merge the arrays
      end
    end
  end
end
```

but instead I need to do this because:
1.  I cannot load the generator files in `tailwindcss-rails` (I think because while they are in the gem, they are not publicly accessible like the rails generators are)
2. Even if I could, the `available_views` method is `private` not `protected` so I can't overwrite it.

Instead, my working solution is to do this instead:

```ruby
# lib/generators/rails/views/views_generator.rb

require 'rails/generators/erb/scaffold/scaffold_generator'
require 'rails/generators/resource_helpers'

module Rails
  module Generators
    # Generate custom views
    class ViewsGenerator < Erb::Generators::ScaffoldGenerator
      include Rails::Generators::ResourceHelpers

      source_root File.expand_path('templates', __dir__)
      source_paths << 'lib/templates/erb/scaffold'

      argument :attributes, type: :array, default: [], banner: 'field[:type][:index] field[:type][:index]'

      def copy_view_files
        %w[_list_item].each do |view|
          formats.each do |format|
            filename = filename_with_extensions(view, format)
            template filename, File.join('app/views', controller_file_path, filename)
          end
        end
      end
    end
  end
end
```

which leaves the original Tailwindcss views generator alone, and only generates my one extra view at the end.

We can close this discussion now, but if anyone has a better approach, I would be very curious to learn it.

**Reply by flavorjones** [reply]
@pas256 Thanks for explaining more.

> I cannot load the generator files in tailwindcss-rails (I think because while they are in the gem, they are not publicly accessible like the rails generators are)

Not true! While they are not _automatically_ loaded, you should still be able to require it:

```ruby
require 'rails/generators'
require 'generators/tailwindcss/scaffold/scaffold_generator'
```

> Even if I could, the available_views method is private not protected so I can't overwrite it.

Not true! In Ruby you can override private methods. You should be able to do this:

```ruby
require 'rails/generators'
require 'generators/tailwindcss/scaffold/scaffold_generator'

class ViewsGenerator < Tailwindcss::Generators::ScaffoldGenerator
  include Rails::Generators::ResourceHelpers

  source_root File.expand_path('templates', __dir__)
  source_paths << 'lib/templates/erb/scaffold'

  argument :attributes, type: :array, default: [], banner: 'field[:type][:index] field[:type][:index]'

  def available_views
    %w(index edit show new _form _list_view)
  end
end
```

**Reply by pas256** [reply]
That is some magic right there. Thank you!

Why is the `require`:

    require 'generators/tailwindcss/scaffold/scaffold_generator'

and not

    require "tailwindcss-rails/generators/tailwindcss/scaffold/scaffold_generator"

Since I am loading a file from a gem, how is it possible that you don't have to specify it's name? ✨ 👻

**Reply by flavorjones** [reply]
Glad it was helpful!

The convention is that a gem's `lib` folder is added to the `require` search path (though this is configurable in the gemspec). So any file in any gem can *usually* be loaded by requiring the path segment below `lib`.

**Comment by walterdavis** [comment]
Rails uses a unified "load path", and a gem will (usually) register its own resources such that Rails can find them by looking them up in the load path. That way you don't need to change so many things per gem you add (or remember to fix that when you remove them later).

**Comment by pas256** [comment]
This has been very helpful. Thank you.
